### ISSUE TITLE
scvi - promote batch bias vectors to latent variables

### ISSUE THREAD

**Issue by sjfleming** [cellarium-ai/cellarium-ml#145]
We would like to have the option to promote batch bias vectors to latent variables.

This amounts to doing two things during inference:

- augment the loss function with a new KL divergence
- learn the distribution of each $d$-dimensional batch $s$ bias vector (for example, $[\mu_{sd}, \sigma_{sd}]$), and sample from that distribution when computing the reconstruction loss according to whatever variational posterior we choose: $W_{sd} \sim \mathcal{N} (\mu_{sd}, \sigma_{sd})$

<img width="413" alt="image" src="https://github.com/cellarium-ai/cellarium-ml/assets/10214815/7f0e41ec-5038-4b3f-b8ce-1eac01fbf3f6">

**Comment by sjfleming** [comment]
The encoder and decoder also need to have additional flexibility added... they need to accommodate this different parameterization.

Don't just have the one-hot concatenation as the only option... allow a batch bias vector to be added.

**Comment by sjfleming** [comment]
Probably want to make a subclass of `torch.nn.Linear` in order to accommodate a learnable bias vector.

**Comment by sjfleming** [comment]
We ultimately want to learn not the bias vectors per batch by themselves, but their _distributions_.  So maybe like a mean and a variance.

**Comment by ImXman** [comment]
I am thinking of a forward function that takes input (x, b). torch.nn.Linear will take x as input to get out1, and an embedding layer takes b to project batch label to out2. The final output would be out=out1+out2.

**Comment by ImXman** [comment]
I looked up the implementation in scvi-tools. The way they implemented batch as embedding is concatenating gene count and batch embedding as an dimension-expanded input. The new input is forwarded through torch.nn.Linear. This is not what we want to implement, right?

**Comment by sjfleming** [comment]
Right, we’d like to do it differently so that the batch bias vector can be sampled from a distribution (and not just computed directly using a weight matrix).  There could be multiple ways to implement this, but I don’t think the current scvi implementation can do what we need

**Comment by sjfleming** [comment]
`LinearWithBatch` now accepts an __init__ argument called `sample`, which is supposed to make this easier.  #156 has been merged to the `modified_scvi` branch, so please base your work on that branch @ImXman
